### PR TITLE
Proto add state

### LINKED ISSUE #6352: Yul proto converter: Add state to the proto-to-yul converter
## Abstract

The current implementation of the yul proto converter (the code that converts a protobuf formatted input to a yul program) does not contain any state.

This means the following are possible
  - referencing as yet undeclared variables
  - nested for loop re-using the same loop variable
  - switch statements with duplicate case literals

Although these issues are to be addressed separately, a fundamental requirement and a pre-requisite to address these problems is to add state to the proto converter.

## Motivation

Adding state requires the following things
  - Create an umbrella class for the converter
  - Add visitor functions to convert protobuf messages
  - Add state to make visits context dependent

## Specification

The `google::protobuf` namespace contains a template class called `RepeatedPtrField` that is used to instantiate custom protobuf messages.

We can define a visitor for this template class, and specialized visitors for individual protobuf message classes.

It will look like this
```
	template <class T>
	void visit(google::protobuf::RepeatedPtrField<T> const& _repeated_field);
	void visit(BinaryOp const&);
	void visit(Block const&);
	void visit(Literal const&);
	void visit(VarRef const&);
	void visit(Expression const&);
	void visit(VarDecl const&);
	void visit(TypedVarDecl const&);
	void visit(UnaryOp const&);
	void visit(AssignmentStatement const&);
	void visit(IfStmt const&);
	void visit(StoreFunc const&);
	void visit(Statement const&);
	void visit(Function const&);
	void visit(ForStmt const&);
	void visit(CaseStmt const&);
	void visit(SwitchStmt const&);
```

These visitor functions are defined inside a class called `protoConverter` that looks like so
```
class protoConverter
{
public:
	protoConverter() {}
	~protoConverter() {}
	std::string functionToString(Function const& _input);
	std::string protoToYul(uint8_t const* _data, size_t _size);

private:
	void visit(BinaryOp const&);
	void visit(Block const&);
```

### PR DESCRIPTION
**Pull request by bshastry**
### Description

(closes #6352 )

### CODE REVIEW

File: test/tools/ossfuzz/protoToYul.h [hunk 1]
```diff
@@ -27,30 +29,40 @@ namespace test
...
-std::ostream& operator<<(std::ostream& _os, IfStmt const& _x);
-std::ostream& operator<<(std::ostream& _os, StoreFunc const& _x);
-std::ostream& operator<<(std::ostream& _os, Statement const& _x);
-std::ostream& operator<<(std::ostream& _os, Block const& _x);
-std::ostream& operator<<(std::ostream& _os, Function const& _x);
-std::ostream& operator<<(std::ostream& _os, ForStmt const& _x);
-std::ostream& operator<<(std::ostream& _os, CaseStmt const& _x);
-std::ostream& operator<<(std::ostream& _os, SwitchStmt const& _x);
-}
+class protoConverter
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.h]: Should be `ProtoConverter`

File: test/tools/ossfuzz/protoToYul.h [hunk 2]
```diff
@@ -27,30 +29,40 @@ namespace test
...
+{
+public:
+	protoConverter() {}
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.h]: Beter to not mention the constuctors and destructors at all.

Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.h]: I will be using them shortly in an upcoming PR that adds scoping to variable declarations. Can we keep it until then, and take a call?

File: test/tools/ossfuzz/yulProtoFuzzer.cpp [hunk 3]
```diff
@@ -30,9 +30,10 @@ using namespace yul;
 using namespace yul::test::yul_fuzzer;
 using namespace std;
 
-DEFINE_BINARY_PROTO_FUZZER(Function const& _input)
+DEFINE_PROTO_FUZZER(Function const& _input)
 {
-	string yul_source = functionToString(_input);
+	protoConverter converter;
```
Comment (erak) [inline test/tools/ossfuzz/yulProtoFuzzer.cpp]: Shouldn't this be `ProtoConverter`?

File: test/tools/ossfuzz/protoToYul.h [hunk 4]
```diff
@@ -27,30 +29,37 @@ namespace test
...
+	void visit(UnaryOp const&);
+	void visit(AssignmentStatement const&);
+	void visit(IfStmt const&);
+	void visit(StoreFunc const&);
+	void visit(Statement const&);
+	void visit(Function const&);
+	void visit(ForStmt const&);
+	void visit(CaseStmt const&);
+	void visit(SwitchStmt const&);
+	template <class T>
```
Comment (erak) [inline test/tools/ossfuzz/protoToYul.h:55]: Maybe add a new line?

File: test/tools/ossfuzz/protoToYul.h [hunk 5]
```diff
@@ -27,30 +29,37 @@ namespace test
...
+	void visit(google::protobuf::RepeatedPtrField<T> const& _repeated_field);
+
+	std::string createHex(std::string const& _hexBytes);
```
Comment (erak) [inline test/tools/ossfuzz/protoToYul.h]:
```suggestion
	std::string createHex(std::string const& _hexBytes) const;
```

Comment (erak) [reply test/tools/ossfuzz/protoToYul.h]: Same could be applied to `createAlphaNum`

**Review by erak — COMMENTED** [review]
Looks good, apart from my tiny comments.

File: test/tools/ossfuzz/protoToYul.h (hunk 4 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.h:55]: Do you mean like so?

```
void visit(SwitchStmt const&);

template <class T>
...
```

**Review by chriseth — APPROVED** [review]

**Pull request merged**